src/modules/time_persistor/TimePersistor.cpp

```diff
@@ -0,0 +1,177 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2024 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "TimePersistor.hpp"
+
+#include <errno.h>
+#include <px4_platform_common/log.h>
+
+TimePersistor::TimePersistor() :
+	ModuleParams(nullptr)
+{
+}
+
+TimePersistor::~TimePersistor()
+{
+	fclose(_file);
+}
+
+int TimePersistor::init()
+{
+	if ((_file = fopen(TIME_FILE_PATH, "r+")) == nullptr) {
+		_file = fopen(TIME_FILE_PATH, "w+");
+	}
+
+	if (!_file) {
+		PX4_ERR("error opening the time file");
+		return PX4_ERROR;
+	}
+
+	time_t t;
+
+	if (read_time(&t) == PX4_OK) {
+		struct timespec ts;
+		ts.tv_sec = t;
+		ts.tv_nsec = 0;
+
+		px4_clock_settime(CLOCK_REALTIME, &ts);
+		return PX4_OK;
+	}
+
+	return PX4_OK;
+}
+
+void TimePersistor::run()
+{
+	while (!should_exit()) {
+		struct timespec ts;
+		px4_clock_gettime(CLOCK_REALTIME, &ts);
+
+		if (write_time(ts.tv_sec) != PX4_OK) {
+			PX4_ERR("error writing RTC to time file");
+			exit_and_cleanup();
+		}
+
+		px4_usleep(1_s);
+	}
+}
+
+int TimePersistor::task_spawn(int argc, char *argv[])
+{
+	_task_id = px4_task_spawn_cmd("time_persistor",
```

Comment:
Memory-wise this would be a bit cheaper running in the PX4 work queue .

ScheduledWorkItem(MODULE_NAME, px4::wq_configurations::lp_default)

Reply:
You could simply call `init()` from `Run()` if the file isn't open.

Reply:
Good idea. Fixed.